modal/shared_volume.py

```diff
@@ -64,6 +66,47 @@ async def read_file(self, path: str) -> AsyncIterator[bytes]:
             async for data in blob_iter(response.data_blob_id, self._client.stub):
                 yield data
 
+    async def add_local_file(
+        self, local_path: Union[Path, str], remote_path: Optional[Union[str, PurePosixPath, None]] = None
+    ):
+        local_path = Path(local_path)
+        if remote_path is None:
+            remote_path = PurePosixPath("/", local_path.name).as_posix()
+        else:
+            remote_path = PurePosixPath(remote_path).as_posix()
+
+        with local_path.open("rb") as local_file:
+            await self.write_file(remote_path, local_file)
+
+    async def add_local_dir(
+        self,
+        local_path: Union[Path, str],
+        remote_path: Optional[Union[str, PurePosixPath, None]] = None,
+        concurrency=20,
+    ):
+        local_path = Path(local_path)
+        if remote_path is None:
+            remote_path = PurePosixPath("/", local_path.name).as_posix()
+        else:
+            remote_path = PurePosixPath(remote_path).as_posix()
+
+        assert local_path.is_dir()
+
+        def gen_transfers():
+            for subpath in local_path.rglob("*"):
+                if subpath.is_dir():
+                    continue
+                relpath_str = subpath.relative_to(local_path).as_posix()
+                yield self.add_local_file(subpath, PurePosixPath(remote_path, relpath_str))
+
+        sem = asyncio.Semaphore(concurrency)
+
+        async def limited_task(coro):
+            async with sem:
+                return await coro
+
+        await asyncio.gather(*(limited_task(t) for t in gen_transfers()), return_exceptions=True)
```

Comment:
minor nit but i think it would be nice to break this out into a general purpose utility function to map with a pool